File: python/semantic_kernel/agents/runtime/in_process/message_handler_context.py
# Copyright (c) Microsoft. All rights reserved.

from collections.abc import Generator
from contextlib import contextmanager
from contextvars import ContextVar
from typing import Any, ClassVar

from semantic_kernel.agents.runtime.core.agent_id import AgentId
from semantic_kernel.utils.feature_stage_decorator import experimental


@experimental
class MessageHandlerContext:
    """Context for message handlers."""

    def __init__(self) -> None:
        """Instantiate the MessageHandlerContext class."""
        raise RuntimeError(
            "MessageHandlerContext cannot be instantiated. It is a static class that provides context management for "
            "message handling."
        )

    _MESSAGE_HANDLER_CONTEXT: ClassVar[ContextVar[AgentId]] = ContextVar("_MESSAGE_HANDLER_CONTEXT")

    @classmethod
    @contextmanager
    def populate_context(cls, ctx: AgentId) -> Generator[None, Any, None]:
        """Populate the context with the current agent ID."""
        token = MessageHandlerContext._MESSAGE_HANDLER_CONTEXT.set(ctx)
        try:
            yield
        finally:
            MessageHandlerContext._MESSAGE_HANDLER_CONTEXT.reset(token)

    @classmethod
    def agent_id(cls) -> AgentId:
        """Get the current agent ID."""
        try:
            return cls._MESSAGE_HANDLER_CONTEXT.get()
        except LookupError as e:
            raise RuntimeError("MessageHandlerContext.agent_id() must be called within a message handler.") from e

File: nodejs/semantic-kernel/agents/runtime/in-process/message-handler-context.ts
import { AsyncLocalStorage } from 'async_hooks'
import { experimental } from '../../../utils/feature-stage-decorator'
import type { AgentId } from '../core/agent-id'

/**
 * Context for message handlers using AsyncLocalStorage to track the current agent handling a message.
 */
@experimental
export class MessageHandlerContext {
  private static _storage = new AsyncLocalStorage<AgentId>()

  /**
   * MessageHandlerContext cannot be instantiated. It is a static class that provides
   * context management for message handling.
   */
  private constructor() {
    throw new Error(
      'MessageHandlerContext cannot be instantiated. It is a static class that provides context management for message handling.'
    )
  }

  /**
   * Populate the context with the current agent ID and execute the callback within that context.
   * @param agentId - The agent ID to set in the context
   * @param callback - The callback to execute within the context
   * @returns The result of the callback
   */
  static populateContext<T>(agentId: AgentId, callback: () => T | Promise<T>): T | Promise<T> {
    return this._storage.run(agentId, callback)
  }

  /**
   * Get the current agent ID from the context.
   * @returns The current agent ID
   * @throws Error if called outside of a message handler context
   */
  static agentId(): AgentId {
    const agentId = this._storage.getStore()
    if (!agentId) {
      throw new Error('MessageHandlerContext.agentId() must be called within a message handler.')
    }
    return agentId
  }
}
